adobe/spectrum-web-components
# feat(calendar): keyboard navigation

## Description

Mainly this PR adds the keyboard navigation functionality to the Calendar's grid of days. As per [WAI keyboard support](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/#kbd_label) for date picker dialogs, the selected day is the only one tabbable from the grid. If there is no day selected the fallback will be the current day and the first day of the displayed month. After the said day is in focus, the following keys and functions apply.

|Key| Action |
| :---:  | :---  |
| Space/ Enter|Selects the focused date|
|Up/ Down Arrow|Moves the focus to the same day of the previous/next available week. An "available week" is a week that has at least one day available. If the said date is disabled, the focus is moved to the closest available day with respect to the navigation direction (a minimum focus movement difference).|
|Left/ Right Arrow|Moves the focus to the previous/next available day. It can go from one week to another if the focus is currently on the first available day of the week|

If the user changes the focus out from the grid and back in, his previous location should be the one tabbable.

Also, this PR addresses some code suggestions/improvements such as:
- removing the unnecessary _"spectrum-Calendar-*"_ from class names
- small performance gains when parsing date table cell properties such as "isToday" or "isSelected"
- passing a single method for every date cell for click events (see previous [comment](https://github.com/adobe/spectrum-web-components/pull/3327#discussion_r1274260137))
- don't dispatch change event when the user selects the previously selected date
- allow clicks only in the date cell radius shown by the styles for consistent UX

While creating tests for this new functionality I also overwritten old tests. Some tests have a _TODO_ comment and will be addressed in the next PR to account for the new value API of the components.

## Related issue(s)
- https://github.com/adobe/spectrum-web-components/issues/2559
- https://github.com/adobe/spectrum-web-components/issues/2560

## Motivation and context

Not having keyboard navigation the Calendar wouldn't be keyboard accessible.

## How has this been tested?
There can be many use-cases and hopefully most of them are covered in unit tests. As per the behaviour explained above, you can apply said key functions by using the component's storybook.

-   [ ] _Test case_
    1. Go to the [Calendar's storybook](https://imizga-calendar-keyboard--spectrum-web-components.netlify.app/storybook/?path=/story/calendar--default)
    2. Click somewhere on the storybook's iframe (where the component is displayed)
    3. Press _Tab_ to navigate through the Previous/Next month buttons as well as the selected/initial keyboard navigable day
    4. Use your arrow keys (_up/down/left/right_) to move through the calendar's days
    5. Use _Space/Enter_ to select the current day
    6. Play around with the mentioned keys, including _Tab_


## Screenshots (if appropriate)

https://github.com/user-attachments/assets/dc998e94-515e-4d1e-b843-c1256d60ece4


## Types of changes




## Checklist




## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.
